feat: Added Exit Process Timer

MeisterRed opened the pull request:
- Added Exit Process Timer
- Added Custom Exit Process Timer (User can specify time while creating a session)
- Updated Session Test 

The exit process timer will cause the plumadriver to shutdown if no requests have been made within a certain time period. The default I have it set to is 2 minutes (120 seconds) but the user can change this when they create a new session. The way this works is when a user makes any request, the timer is reset to 0 and will start incrementing till it reaches the specified threshold (the default being 120 seconds like I said earlier).

If there's any changes that need to be made or any possible errors that I missed, I'll fix them as soon as possible.

closes #108

pynnl commented on src/Session/Session.ts:
```diff
@@ -187,6 +197,7 @@ class Session {
       'pageLoadStrategy',
       'proxy',
       'timeouts',
+      'idleTime',
```
Firstly, if the timer is an option when creating sessions, it should be separated per session. If I see it correctly, your current implementation of `timer` is shared across sessions. Also as we discussed with @ctyler on Tuesday, you might want to consider it as an command line argument instead, for simplicity.

Secondly, this option is not specified by W3C, so you might want to put it in `plm:plumaOptions`.

MeisterRed replied:
The main goal was to have the timer shared within all sessions. The issue never specified a timer for each separate session, just that the driver will quit after a certain time. It also suggested that the user can set the idle duration while creating a session as well which is what I was trying to go for here. 

I'll make the adjustments and move it to `plm:plumaOptions`

pynnl replied:
The issue is there for discussion, you can always give suggestions to improve it. Although it doesn't tell exactly, but the options when creating a new session are usually bound to that session. It doesn't make sense if I create session `A`, then later on I create session `B` which will change some settings in session `A`. If it's a global setting, then it should be a separated endpoint and won't be relating to session creation (such as a command line argument as suggested above).

pynnl commented on src/Session/command-handlers/add-cookie.ts:
```diff
@@ -1,6 +1,8 @@
+import { updateTimer } from '../../timer';
 
 const addCookie: Pluma.CommandHandler = async ({ session, parameters }) => {
+  updateTimer();
```
Instead of updating the timer inside every commands, maybe doing that in an entry level which having all of the commands?

pynnl commented on src/index.ts:
```diff
@@ -1,9 +1,12 @@
 app.listen(HTTP_PORT, () => {
+  updateTimer();
   console.info(`plumadriver is listening on port ${HTTP_PORT}`);
+  setInterval(idleShutdown, 10000);
```
This method will check the timer every 10s, so it won't be very accurate (for example with `idleTime = 11`, it's supposed to timeout after 11s of inactivity, but instead it'll be after 20s). You can lower the interval, but then the timer has to check more frequently. So, I suggest using `setTimeout` for this (`timer` logic need to be changed as well).

ctyler replied:
Generally, you want to avoid polling whenever possible. Instead, cancel and re-set the timeout on each update (which is a variation on the standard 'watchdog timer' design pattern).

pynnl reviewed (commented):
Some quick reviews so you can continue your works.

MeisterRed commented:
I've made changes based on the feedback I've received from this pull request. The implementation now uses `setTimeout` instead of `setInterval` to avoid polling. I've also put `idleTime` in PlumaOptions and simplified the implementation so the timer is updated at an entry level instead of all of the commands.

The exit process timer still functions the same way it did before. This implementation is simple enough to get the job done but can be improved in a future github issue.

pynnl commented on src/Types/types.d.ts:
```diff
@@ -128,6 +130,7 @@ declare namespace Pluma {
     browserVersion: string;
     platformName: NodeJS.Platform;
     setWindowRect: boolean;
+    idleTime?: number;
```
Is this redundant, cause we already have `idleTime` in `plumaOptions`?

pynnl commented on test/jest/e2e/session.test.js:
```diff
@@ -25,6 +25,7 @@ describe('Session', () => {
             pageLoad: 200,
             script: 300,
           },
+          idleTime: 60,
```
Should be in `plumaOptions` cause this is not w3c standard

pynnl commented on test/jest/e2e/session.test.js:
```diff
@@ -44,6 +45,7 @@ describe('Session', () => {
             pageLoad: 200,
             script: 300,
           },
+          idleTime: 60,
```
Same here

pynnl commented on src/SessionManager/SessionManager.ts:
```diff
@@ -55,6 +55,7 @@ class SessionManager {
             session.browser.browserConfig.unhandledPromptBehavior,
           proxy: session.proxy ? session.proxy : {},
           timeouts: session.timeouts,
+          idleTime: session.browser.browserConfig.idleTime,
```
If I get it correctly, the logic is handled at router layer. If so, there's no need to pass it to session layer. I see it's not being used anywhere inside session.

MeisterRed replied:
The logic for updating the idle timeout threshold is done during the request of creating a new session. this is handled in the `confirgureSession()` function and it uses this `idleTime`

pynnl replied:
Ah I see. Thanks for clarifying.

pynnl commented on src/Types/types.d.ts:
```diff
@@ -26,6 +26,7 @@ declare namespace Pluma {
     strictSSL: boolean;
     unhandledPromptBehavior: unhandledPromptBehavior;
     rejectPublicSuffixes: boolean;
+    idleTime: number;
```
Same here, this could be unused.

pynnl commented on src/routes/index.ts:
```diff
@@ -18,8 +19,27 @@ const {
+  idleTimeout = setTimeout(exit, idleTime);
+  next();
+}
+
```
Just a suggestion, since you already created a `timer.ts`, maybe it's better to make a `IdleTimeoutManager` that holds all the logic? That might help make the code base more organized.

pynnl reviewed (commented):
Just some minor suggestions. Also, please make a test to make sure the feature functions as it should.

poftadeh commented on src/Browser/BrowserConfig.ts:
```diff
@@ -15,6 +15,9 @@ export default class BrowserConfig {
   strictSSL = true;
 
+  /** the time set where plumadriver will automatically close when there's no activity */
+  idleTime = 120;
```
nit: I would suggest a slightly more specific name such as `maxIdleTime`.

MeisterRed replied:
Sure, `maxIdleTime` would be a much better name

poftadeh commented on src/routes/index.ts:
```diff
@@ -18,8 +19,27 @@ const {
+  process.exit(0);
+}
+
+let idleTimeout = setTimeout(exit, 120000);
```
Does this need to be initialized? `clearTimeout(undefined)` would not throw.

MeisterRed replied:
I originally thought it would throw since I was having errors before but it turns out you're right. I guess the errors I had before were from something else. I changed this line of code to remove the initialization

poftadeh commented on src/routes/index.ts:
```diff
@@ -42,12 +62,16 @@ router.post('/session', async (req, res, next) => {
   }
 });
 
+router.use(setIdleTimeout);
```
Correct me if I'm wrong, but idle timeout is on by default? It should be opt-in by the user since this is not typical webdriver behaviour.

MeisterRed replied:
Would changing the default time from 2 minutes to a very high number (an abnormal threshold) be fine?

MeisterRed replied:
Never mind, I managed to come up with a better solution

poftadeh commented on test/jest/e2e/session.test.js:
```diff
@@ -25,6 +25,7 @@ describe('Session', () => {
             pageLoad: 200,
             script: 300,
           },
```
Please add a test for `exit` being called 🙂 . You can use a combination of jest's `spyOn` and `mockImplementation` on `process.exit` to prevent it from actually exiting in the middle of a test. The `toHaveBeenCalled` assertion can check if the function was called after timeout.

poftadeh reviewed (commented):
Added some comments, but it's looking good so far.

MeisterRed commented:
I've made the requested changes and created a test for the idle timeout feature. Please let me know if you have any questions or have any suggestions on what should be changed.

pynnl commented on test/jest/e2e/helpers.js:
```diff
@@ -3,13 +3,21 @@ exports.createSession = async (request, app) => {
+            runScripts: true,
+            idleTimer: true,
+            maxIdleTime: 4
+          },
```
The `helper` is used by other tests, so you might not wanna enable `idleTimer` here since it's disabled by default. I suggest doing similar to `session.test.js`

pynnl commented on test/jest/e2e/shutdown.test.js:
```diff
@@ -21,4 +22,35 @@ describe('Shutdown', () => {
   });
+
+  it('exits a PlumaDriver process after 10 seconds', async () => {
+    await createSession(request, app);
```
This might not be needed as you will call `createSession` down below

pynnl commented on test/jest/e2e/shutdown.test.js:
```diff
@@ -21,4 +22,35 @@ describe('Shutdown', () => {
+    .spyOn(process, 'exit')
+    .mockImplementation(() => (isProcessTerminated = true));
+
+    setInterval(()=>{
```
Please clear the interval after the test

pynnl commented on test/jest/e2e/shutdown.test.js:
```diff
@@ -21,4 +22,35 @@ describe('Shutdown', () => {
+
+    expect(createFn).toHaveBeenCalledTimes(2);
+
+    await waitForExpect(()=> {
```
I believe this test can be done without `waitForExpect`. With `maxIdleTimer = 4s`, your last `createSession` call is 3s after the first call, so it's supposed to exit after 7s. Now you can just advance the timer by 8s, then test if `isProcessTerminated` is true.

Besides, the current test only checks if process is terminated at the end, but doesn't check if process is NOT terminated within the first 3s. You should also have a check in your `interval`.

MeisterRed replied:
`jest.advanceTimersByTime()` seems to only affect timeouts and Intervals created within the test. I've tried many times and couldn't get time to advance for the idle timeout itself. I used `waitForExpect` so that the test doesn't end until the process ends at the requested time.

pynnl replied:
I see. Thanks for clarifying.

pynnl reviewed (commented):
Looking great. Just some comments.

MeisterRed commented:
I've updated the tests based on the suggestions given to me. I moved the idle timeout tests into it's own sperate jest file and I've split it into two tests: one that checks if the process is ever ended earlier than expected, and one that checks if the process ends on time.

The "exits a PlumaDriver process after 3 seconds" test works the way it originally did but with less code. a session will be created with the timeout being 3 seconds and after the 3 seconds are over, we will check if the process has actually ended.

The "check if process exits early before being reinitialized" test works similarly but the check condition is different. First, a session is created with the idle timeout being 5 seconds. Sometime after that, another session will be created changing the idle timeout to 10 seconds. After 7 seconds have passed, it will then check if the process has ended based on the original timeout instead of the new initialized timeout. If `isProcessTerminatedEarly` is still false then the test is a success

pynnl commented on test/jest/e2e/helpers.js:
```diff
@@ -34,4 +32,4 @@ exports.createSession = async (request, app) => {
 
   return sessionId;
-};
+};
```
Newline at eof

pynnl commented on test/jest/e2e/idle-timeout.test.js:
```diff
@@ -0,0 +1,74 @@
+    }, 7000)
+    jest.runAllTimers();
+  });
+});
```
Newline at eof. Also there's a few formatting issues such as indent or space between brackets. Right now we don't have yet eslint check for `.js` files, so I suggest using vscode auto format for this.

pynnl reviewed (changes requested):
Thanks for making the changes. LGTM, just some formatting issues.

MeisterRed commented:
Alright, I fixed up the formatting issues you have pointed out
